pkg/utils/revision/revision_utils.go

```diff
@@ -0,0 +1,301 @@
+	"k8s.io/apimachinery/pkg/util/rand"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	leaderworkerset "sigs.k8s.io/lws/api/leaderworkerset/v1"
+)
+
+// Functions in this package are adapted from https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/ and
+// https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/history/controller_history.go
+
+// EqualRevision returns true if lhs and rhs are either both nil, if the templateRevisionHash is the same,
+// or if they are semantically equivalent.
+func EqualRevision(lhs *appsv1.ControllerRevision, rhs *appsv1.ControllerRevision) bool {
+	if lhs == nil || rhs == nil {
+		return lhs == rhs
+	}
+
+	if GetRevisionKey(lhs) == GetRevisionKey(rhs) {
+		return true
+	}
+
+	return bytes.Equal(lhs.Data.Raw, rhs.Data.Raw) && apiequality.Semantic.DeepEqual(lhs.Data.Object, rhs.Data.Object)
+}
+
+// ListRevisions lists all ControllerRevisions matching selector and owned by parent or no other
+// controller. If the returned error is nil the returned slice of ControllerRevisions is valid. If the
+// returned error is not nil, the returned slice is not valid.
+func ListRevisions(ctx context.Context, k8sClient client.Client, parent metav1.Object, selector labels.Selector) ([]*appsv1.ControllerRevision, error) {
+	// List all revisions in the namespace that match the selector
+	revisionList := new(appsv1.ControllerRevisionList)
+	err := k8sClient.List(ctx, revisionList, client.InNamespace(parent.GetNamespace()), client.MatchingLabelsSelector{Selector: selector})
+	if err != nil {
+		return nil, err
+	}
+	history := revisionList.Items
+	var owned []*appsv1.ControllerRevision
+	for i := range history {
+		ref := metav1.GetControllerOfNoCopy(&history[i])
+		if ref == nil || ref.UID == parent.GetUID() {
+			owned = append(owned, &history[i])
+		}
+
+	}
+	return owned, err
+}
+
+func GetRevision(ctx context.Context, k8sClient client.Client, lws *leaderworkerset.LeaderWorkerSet, templateHash string) (*appsv1.ControllerRevision, error) {
+	log := ctrl.LoggerFrom(ctx).WithValues("leaderworkerset", klog.KObj(lws))
+	ctx = ctrl.LoggerInto(ctx, log)
+	if templateHash == "" {
+		return nil, nil
+	}
+	selector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{MatchLabels: map[string]string{
+		leaderworkerset.RevisionKey: templateHash,
+	}})
+	if err != nil {
+		return nil, err
+	}
+	revisions, err := ListRevisions(ctx, k8sClient, lws, selector)
+	if err != nil {
+		log.Error(err, "Listing all controller revisions")
+		return nil, err
+	}
+
+	if len(revisions) == 0 {
+		return nil, nil
+	}
+
+	if len(revisions) > 1 {
+		// Since we only create a controllerRevision when the template hash changes, only one should match
+		log.Error(err, "More than one revision exists for the given template hash; returning the latest revision")
+		return revisions[len(revisions)-1], nil
```

Comment:
@ahg-g is there a guarantee that the latest is the one it the highest index? The revisions are not sorted

Reply:
no, we need to sort here.

Reply:
Instead of sorting, wdyt of iterating through the list and returning the one with highest revision value?

Reply:
oh, yeah, if that is all we need

Reply:
change the semantics of `nextRevision` to `getHighestRevision`, use it here and in NewRevision (where you would increment by one to get the next revision